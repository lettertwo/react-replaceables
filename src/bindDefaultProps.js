import {PropTypes} from 'react';
import compose from 'recompose/compose';
import getContext from 'recompose/getContext';
import defaultProps from 'recompose/defaultProps';
import componentFromProp from 'recompose/componentFromProp';

const replacedContext = {replacedComponent: PropTypes.func.isRequired};


export default function bindDefaultProps(props) {
  const enhance = compose(defaultProps(props), getContext(replacedContext));
  return enhance(componentFromProp('replacedComponent'));
}
