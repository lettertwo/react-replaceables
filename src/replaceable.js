import React, {PropTypes} from 'react';
import invariant from 'invariant';


export default function replaceable(Component) {
  if (Component === undefined) return (Wrapped = null) => replaceable(Wrapped);

  invariant(
    typeof Component === 'function',
    'Only React Component classes or stateless function components can be replaceable!'
  );

  const displayName = Component.displayName || Component.name;

  invariant(displayName,
    'Cannot replace an anonymous component!' +
    ' Give your statelss function component a name or displayName property.'
  );

  return React.createClass({
    displayName: `${displayName} (replaceable)`,
    contextTypes: {componentReplacements: PropTypes.object},
    render() {
      const {componentReplacements: replacements = {}} = this.context;
      const {[displayName]: ReplacementComponent = Component} = replacements;
      return <ReplacementComponent {...this.props} />;
    },
  });

}
