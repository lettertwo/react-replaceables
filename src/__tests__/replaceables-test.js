/* eslint-env jest */
import React, {Component} from 'react';
import renderer from 'react-test-renderer';
import {default as replaceableActual} from '../replaceable';
import {default as bindDefaultPropsActual} from '../bindDefaultProps';
import {
  default as ReplacementActual,
  createReplacement as createReplacementActual,
} from '../replacement';
import Replacement, {
  replaceable,
  bindDefaultProps,
  createReplacement,
  Replacement as ReplacementNamed,
} from '../index';

describe('replaceables', () => {
  it('exports the Replacement component as default', () => {
    expect(Replacement).toBe(ReplacementActual);
  });

  it('exports the Replacement component', () => {
    expect(ReplacementNamed).toBe(ReplacementActual);
  });

  it('exports the replaceable decorator', () => {
    expect(replaceable).toBe(replaceableActual);
  });

  it('exports the bindDefaultProps util', () => {
    expect(bindDefaultProps).toBe(bindDefaultPropsActual);
  });

  it('exports the createReplacement factory', () => {
    expect(createReplacement).toBe(createReplacementActual);
  });

  describe('in a rendered component', () => {
    const Replaceable = replaceable(function Replaceable() {
      return <span>replaceme</span>;
    });

    it('shallowly replaces components', () => {
      const tree = renderer
        .create(
          <Replacement Replaceable={() => <span>test</span>}>
            <Replaceable />
          </Replacement>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('deeply replaces components', () => {
      const tree = renderer
        .create(
          <Replacement Replaceable={() => <span>test</span>}>
            <div>
              <Replaceable />
              <span>decoy</span>
            </div>
          </Replacement>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
