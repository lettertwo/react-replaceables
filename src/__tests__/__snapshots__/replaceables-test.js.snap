// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replaceables in a rendered component deeply replaces components 1`] = `
<div>
  <span>
    test
  </span>
  <span>
    decoy
  </span>
</div>
`;

exports[`replaceables in a rendered component shallowly replaces components 1`] = `
<span>
  test
</span>
`;
